Source/NSubstitute/Core/CallCollection.cs

```diff
@@ -0,0 +1,56 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NSubstitute.Core
+{
+    public class CallCollection : ICallCollection, IReceivedCalls
+    {
+        ConcurrentQueue<CallWrapper> _callWrappers = new ConcurrentQueue<CallWrapper>();
+
+        public void Add(ICall call)
+        {
+            if (call == null) throw new ArgumentNullException(nameof(call));
+            _callWrappers.Enqueue(new CallWrapper(call));
+        }
+
+        public void Delete(ICall call)
+        {
+            if (call == null) throw new ArgumentNullException(nameof(call));
+
+            var callWrapper = _callWrappers.FirstOrDefault(w => w.Call.Equals(call));
+            if (callWrapper == null) throw new InvalidOperationException("Collection doesn't contain the call.");
+
+            callWrapper.IsDeleted = true;
```

Comment:
@dtchepak I can think of the following test (a lot of heavy specs):
```csharp
[Test]
[Ignore]
public void Call_between_invocation_and_received_doesnt_cause_issue()
{
    var subs = Substitute.For<ISomething>();
    for (int i = 0; i < 1000000; i++)
    {
        subs.Echo(0).Returns(new string('a', 100000) + i);
    }
}
```

Do you have a different case?